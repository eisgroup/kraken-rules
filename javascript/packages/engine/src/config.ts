/*
 *  Copyright 2020 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { debug } from './debugger/Debugger'

/**
 * Configures Kraken evaluation. It can be configured from the console.
 * Open the console and  modify the config options
 * @example
 * global.Kraken.debug = false;
 * global.Kraken.info = false;
 * global.Kraken.debug = false;
 * @see Logger
 * @since 1.1.1
 */
export interface KrakenConfig {
    logger: {
        enabled: boolean
        debug: boolean
        info: boolean
        error: boolean
        warning: boolean
        entriesCount: number
        logs: string[]
        getLogs(): void
        clear(): void
    }
    debugger: debug.api.Debugger
}
export type WithKraken = {
    Kraken: KrakenConfig
}

function init(): void {
    const logger: KrakenConfig['logger'] = {
        enabled: process.env.NODE_ENV !== 'production',
        debug: false,
        info: true,
        error: true,
        warning: true,
        entriesCount: 2000,
        logs: [],
        clear(): void {
            const config = (global as unknown as WithKraken).Kraken
            if (config.logger.logs) {
                config.logger.logs = []
            }
        },
        getLogs(): void {
            const config = (global as unknown as WithKraken).Kraken
            // applicable only for the browser
            if (config.logger.logs && document && window) {
                const blob = new Blob([config.logger.logs.join('\n')], { type: 'text/plain' })
                const url = window.URL.createObjectURL(blob)
                const a = document.createElement('a')
                a.style.display = 'none'
                a.href = url
                a.download = `kraken-logs-${new Date().getTime()}.txt`
                document.body.appendChild(a)
                a.click()
                window.URL.revokeObjectURL(url)
                document.body.removeChild(a)
            }
        },
    }
    if (global) {
        ;(global as unknown as WithKraken).Kraken = {
            logger,
            debugger: new debug.impl.DevToolsDebugger(),
        }
        return
    }
}

function getConfig(): KrakenConfig {
    if (global) {
        return (global as unknown as { Kraken: KrakenConfig })['Kraken']
    }
    throw new Error('Kraken configuration was not initialized')
}

export const krakenConfig = { init, getConfig }
