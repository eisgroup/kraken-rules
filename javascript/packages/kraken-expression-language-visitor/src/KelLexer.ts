import { Common } from './antlr/generated/Common'

export class KelLexer extends Common {
    /** @override */
    popMode(): number {
        return this._modeStack.isEmpty ? Common.DEFAULT_MODE : super.popMode()
    }
}
