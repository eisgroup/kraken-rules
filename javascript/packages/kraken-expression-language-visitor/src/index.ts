export { KelVisitor } from './antlr/generated/KelVisitor'
export { KelParser, SyntaxError } from './KelParser'
export * from './antlr/generated/Kel'

export { ParserRuleContext } from 'antlr4ts'
export { AbstractParseTreeVisitor } from 'antlr4ts/tree/AbstractParseTreeVisitor'
export { ErrorNode } from 'antlr4ts/tree/ErrorNode'
export { Interval } from 'antlr4ts/misc/Interval'
export { TerminalNode } from 'antlr4ts/tree/TerminalNode'
