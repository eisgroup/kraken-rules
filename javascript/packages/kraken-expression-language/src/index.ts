export * from './scope/Scope'
export * from './scope/ScopeType'
export * from './visitor/Node'
export * from './visitor/AstGeneratingVisitor'
export * from './visitor/NodeUtils'

export { KelParser, SyntaxError } from 'kraken-expression-language-visitor'

export * from './visitor/autocomplete/CompletionItem'
export * from './visitor/type-at-location/LocationInfo'

export * from './public/Cache'
export * from './public/KelService'
export * from './public/ScopeDeserializer'

export { Type } from './type/Types'
