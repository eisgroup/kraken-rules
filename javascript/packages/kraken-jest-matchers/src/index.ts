/*
 *  Copyright 2022 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { matchers as dateTimeMatchers } from './dateTimeMatchers'
import { matchers as ruleEvalResultMatchers } from './ruleEvalResultMatchers'
import { matchers as entryPointResultMatchers } from './entrypointResultMatchers'

export const matchers = {
    ...dateTimeMatchers,
    ...ruleEvalResultMatchers,
    ...entryPointResultMatchers,
}

export { ExpectedResult } from './entrypointResultMatchers'
export { MatchableResults } from './entrypointResultMatchers'

expect.extend(matchers)
