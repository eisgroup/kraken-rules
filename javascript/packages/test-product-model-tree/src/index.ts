export * from './generated/kraken_model_tree_Policy'
export * from './generated/kraken_model_tree_PolicyExtended'
export * from './generated/kraken_functions_Policy'
export * from './generated/kraken_functions_PolicyExtended'
