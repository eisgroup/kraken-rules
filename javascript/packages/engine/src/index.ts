/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
import { krakenConfig } from './config'
krakenConfig.init()

// spi
export * from './engine/contexts/info/DataObjectInfoResolver'
export * from './engine/contexts/info/ContextInstanceInfoResolver'

// engine
export * from './engine/executer/SyncEngine'

export * from './bundle-cache/EntryPointBundleCache'
export * from './bundle-cache/delta-cache/RepoClientCache'
export * from './bundle-cache/delta-cache/DeltaBundleCache'
export * from './bundle-cache/dimension-set-cache/DimensionSetBundleCache'
export * from './bundle-cache/dimension-set-cache/DimensionSetCacheLogger'
export * from './bundle-cache/expression-context-manager/ExpressionContextManager'
export * from './bundle-cache/expression-context-manager/ExpressionContextManagerImpl'

export { registry, KelFunction, FunctionParameter } from './engine/runtime/expressions/ExpressionEvaluator'
export { FunctionScope } from './engine/runtime/expressions/functionLibrary/Registry'
export * from './engine/runtime/EvaluationMode'

// Payload result utils
export * from './engine/results/PayloadResultCreator'
export * from './engine/results/PayloadResultTypeChecker'
export { ValueChangedEvent } from './engine/results/ValueChangedEvent'

// Condition result utils
export { conditionEvaluationTypeChecker } from './dto/DefaultConditionEvaluationResult'

export { DataContext } from './engine/contexts/data/DataContext'

// field metadata reducer
export * from './engine/results/field_metadata_reducer/FieldErrorMessage'

export * from './models/EntryPointBundle'
export * from './models/ContextModelTree'

// reducer
export * from './engine/results/RuleOverrideContextExtractor'
export * from './engine/results/Reducer'
export * from './engine/results/Localization'

// field metadata reducer
export * from './engine/results/field_metadata_reducer/FieldMetadataReducer'

// error
export * from './error/KrakenRuntimeError'

// re-export
export {
    ContextInstanceInfo,
    RuleOverride,
    RuleInfo,
    ContextFieldInfo,
    EntryPointResult,
    ConditionEvaluationResult,
    FieldEvaluationResult,
    RuleEvaluationResults,
    PayloadResultType,
    PayloadResult,
    ErrorAwarePayloadResult,
    ErrorMessage,
    ValidationPayloadResult,
    AccessibilityPayloadResult,
    VisibilityPayloadResult,
    DefaultValuePayloadResult,
    AssertionPayloadResult,
    SizePayloadResult,
    SizeRangePayloadResult,
    LengthPayloadResult,
    RegExpPayloadResult,
    UsagePayloadResult,
    ValueListPayloadResult,
    RuleEvent,
    FieldMetadata,
    FieldMetadataResult,
    ConditionEvaluation,
} from 'kraken-engine-api'
