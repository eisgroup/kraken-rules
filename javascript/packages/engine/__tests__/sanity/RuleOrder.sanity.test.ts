/* eslint-disable @typescript-eslint/no-non-null-assertion */
/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { sanityMocks } from './_AutoPolicyObject.mocks'
import { sanityEngine } from './_SanityEngine'
import { mock } from '../mock'
import { FieldMetadataReducer } from '../../src'

describe('Engine Sanity rule order test', () => {
    const { empty } = sanityMocks
    it('should execute rules in order', () => {
        const policy = empty()
        policy.riskItems = [
            {
                cd: mock.modelTreeJson.contexts.Vehicle.name,
                id: 'v1',
                collCoverages: [
                    {
                        cd: mock.modelTreeJson.contexts.COLLCoverage.name,
                        id: 'cc1',
                    },
                ],
            },
        ]
        policy.insured = {
            cd: mock.modelTreeJson.contexts.Insured.name,
            id: 'i1',
            haveChildren: true,
        }
        policy.billingInfo = {
            creditCardInfo: {
                cd: mock.modelTreeJson.contexts.CreditCardInfo.name,
                id: 'cci1',
                billingAddress: {
                    cd: mock.modelTreeJson.contexts.BillingAddress.name,
                    id: 'ba1',
                },
            },
        }
        policy.parties = [
            {
                cd: mock.modelTreeJson.contexts.Party.name,
                id: 'p1',
            },
        ]
        const results = sanityEngine.evaluate(policy, 'RuleOrder')
        const fieldResults = new FieldMetadataReducer().reduce(results)

        expect(policy.riskItems[0].collCoverages![0].code).toBe('PartyPartyAddress')
        expect(results).k_toMatchResultsStats({ total: 7, critical: 0, warning: 0 })
        expect(results).k_toHaveExpressionsFailures(0)
        expect({ entryPointResults: results, fieldResults }).k_toMatchResultsSnapshots()
    })
    it('should execute rules with complex field expressions in order', () => {
        const policy = empty()
        policy.billingInfo = {
            creditCardInfo: {
                cd: mock.modelTreeJson.contexts.CreditCardInfo.name,
                id: 'cci1',
            },
        }
        policy.riskItems = [
            {
                cd: mock.modelTreeJson.contexts.Vehicle.name,
                id: 'v1',
                rentalCoverage: {
                    cd: mock.modelTreeJson.contexts.RRCoverage.name,
                    id: 'rr1',
                },
            },
        ]
        policy.insured = {
            cd: mock.modelTreeJson.contexts.Insured.name,
            id: 'i1',
            addressInfo: {
                cd: mock.modelTreeJson.contexts.BillingAddress.name,
                id: 'adi1',
            },
        }
        policy.parties = [
            {
                cd: mock.modelTreeJson.contexts.Party.name,
                id: 'p1',
                roles: [
                    {
                        cd: mock.modelTreeJson.contexts.PartyRole.name,
                        id: 'pr1',
                    },
                ],
            },
        ]
        const results = sanityEngine.evaluate(policy, 'RuleOrderWithComplexField')
        const fieldResults = new FieldMetadataReducer().reduce(results)

        expect(policy.policyNumber).toBe('doNotSolicit is true')
        expect(policy.insured.addressInfo!.doNotSolicit).toBeTruthy()
        expect(policy.riskItems[0].rentalCoverage!.limitAmount).toBe(10)
        expect(policy.insured.addressInfo!.city).toBe('San Diego')
        expect(policy.parties[0].roles![0].limit).toBe(100)
        expect(results).k_toMatchResultsStats({ total: 6 })
        expect({ entryPointResults: results, fieldResults }).k_toMatchResultsSnapshots()
    })
    it('should set default coverage and transaction detail limit amounts to 1000', () => {
        const policy = empty()
        policy.transactionDetails = { id: '2', cd: 'TransactionDetails' }
        policy.coverage = {
            id: '1',
            cd: 'CarCoverage',
        }

        const results = sanityEngine.evaluate(policy, 'FunctionCheck-RulesUsingFunctionOrderCheck')

        expect(policy.coverage.limitAmount).toBe(1000)
        expect(policy.transactionDetails.totalLimit).toBe(1000)

        expect(results).k_toMatchResultsStats({ total: 2, critical: 0 })
        expect(results).k_toHaveExpressionsFailures(0)
    })
    it('shouldEvaluateRuleOrderWithNoCycleInFunction', async () => {
        const policy = empty()
        policy.riskItems = [
            {
                cd: mock.modelTreeJson.contexts.Vehicle.name,
                id: 'v1',
                rentalCoverage: {
                    cd: mock.modelTreeJson.contexts.RRCoverage.name,
                    id: 'rr1',
                },
                collCoverages: [
                    {
                        cd: mock.modelTreeJson.contexts.COLLCoverage.name,
                        id: 'coll1',
                        limitAmount: 50,
                    },
                    {
                        cd: mock.modelTreeJson.contexts.COLLCoverage.name,
                        id: 'coll2',
                        limitAmount: 75,
                    },
                ],
            },
        ]

        const results = sanityEngine.evaluate(policy, 'RuleOrderWithNoCycleInFunction')

        expect(policy.policyNumber).toBe('Large')
        expect(policy.riskItems[0].rentalCoverage!.limitAmount).toBe(125)

        expect(results).k_toMatchResultsStats({ total: 2, critical: 0 })
        expect(results).k_toHaveExpressionsFailures(0)
    })
})
