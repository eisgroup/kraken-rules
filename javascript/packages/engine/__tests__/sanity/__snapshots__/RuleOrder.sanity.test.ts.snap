// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Engine Sanity rule order test should execute rules in order 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "BillingAddress:ba1:countryCd": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "ba1",
          "contextName": "BillingAddress",
          "fieldName": "countryCd",
          "fieldPath": "countryCd",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "countryCd",
                  "contextId": "ba1",
                  "contextName": "BillingAddress",
                  "newValue": "Address",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "AddressInfo",
              "payloadtype": "DEFAULT",
              "ruleName": "RULEORDER-BILLINGADDRESS",
              "targetPath": "countryCd",
            },
          },
        ],
      },
      "COLLCoverage:cc1:code": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "cc1",
          "contextName": "COLLCoverage",
          "fieldName": "code",
          "fieldPath": "code",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "code",
                  "contextId": "cc1",
                  "contextName": "COLLCoverage",
                  "newValue": "PartyPartyAddress",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "COLLCoverage",
              "payloadtype": "DEFAULT",
              "ruleName": "RULEORDER-COLL",
              "targetPath": "code",
            },
          },
        ],
      },
      "CreditCardInfo:cci1:cardType": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "cci1",
          "contextName": "CreditCardInfo",
          "fieldName": "cardType",
          "fieldPath": "cardType",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "cardType",
                  "contextId": "cci1",
                  "contextName": "CreditCardInfo",
                  "newValue": "PartyAddress",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "CreditCardInfo",
              "payloadtype": "DEFAULT",
              "ruleName": "RULEORDER-CREDITCARD",
              "targetPath": "cardType",
            },
          },
        ],
      },
      "Insured:i1:name": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "i1",
          "contextName": "Insured",
          "fieldName": "name",
          "fieldPath": "name",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "name",
                  "contextId": "i1",
                  "contextName": "Insured",
                  "newValue": "Party",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Insured",
              "payloadtype": "DEFAULT",
              "ruleName": "RULEORDER-INSURED",
              "targetPath": "name",
            },
          },
        ],
      },
      "Party:p1:relationToPrimaryInsured": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "p1",
          "contextName": "Party",
          "fieldName": "relationToPrimaryInsured",
          "fieldPath": "relationToPrimaryInsured",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "relationToPrimaryInsured",
                  "contextId": "p1",
                  "contextName": "Party",
                  "newValue": "Party",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Party",
              "payloadtype": "DEFAULT",
              "ruleName": "RULEORDER-PARTY",
              "targetPath": "relationToPrimaryInsured",
            },
          },
        ],
      },
      "Policy:0:policyNumber": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "policyNumber",
          "fieldPath": "policyNumber",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "policyNumber",
                  "contextId": "0",
                  "contextName": "Policy",
                  "newValue": "PartyPartyAddress",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "DEFAULT",
              "ruleName": "RULEORDER-POLICY",
              "targetPath": "policyNumber",
            },
          },
        ],
      },
      "Vehicle:v1:model": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "v1",
          "contextName": "Vehicle",
          "fieldName": "model",
          "fieldPath": "model",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "model",
                  "contextId": "v1",
                  "contextName": "Vehicle",
                  "newValue": "PartyPartyAddress",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Vehicle",
              "payloadtype": "DEFAULT",
              "ruleName": "RULEORDER-VEHICLE",
              "targetPath": "model",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "BillingAddress:ba1:countryCd": Object {
      "fieldType": "STRING",
      "id": "BillingAddress:ba1:countryCd",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "countryCd",
      "ruleResults": Array [],
    },
    "COLLCoverage:cc1:code": Object {
      "fieldType": "STRING",
      "id": "COLLCoverage:cc1:code",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "code",
      "ruleResults": Array [],
    },
    "CreditCardInfo:cci1:cardType": Object {
      "fieldType": "STRING",
      "id": "CreditCardInfo:cci1:cardType",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "cardType",
      "ruleResults": Array [],
    },
    "Insured:i1:name": Object {
      "fieldType": "STRING",
      "id": "Insured:i1:name",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "name",
      "ruleResults": Array [],
    },
    "Party:p1:relationToPrimaryInsured": Object {
      "fieldType": "STRING",
      "id": "Party:p1:relationToPrimaryInsured",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "relationToPrimaryInsured",
      "ruleResults": Array [],
    },
    "Policy:0:policyNumber": Object {
      "fieldType": "STRING",
      "id": "Policy:0:policyNumber",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyNumber",
      "ruleResults": Array [],
    },
    "Vehicle:v1:model": Object {
      "fieldType": "STRING",
      "id": "Vehicle:v1:model",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "model",
      "ruleResults": Array [],
    },
  },
}
`;

exports[`Engine Sanity rule order test should execute rules with complex field expressions in order 1`] = `
Object {
  "entryPointResults": Object {
    "evaluationTimestamp": Any<Date>,
    "results": Object {
      "BillingAddress:adi1:city": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "adi1",
          "contextName": "BillingAddress",
          "fieldName": "city",
          "fieldPath": "city",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "city",
                  "contextId": "adi1",
                  "contextName": "BillingAddress",
                  "newValue": "San Diego",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "AddressInfo",
              "payloadtype": "DEFAULT",
              "ruleName": "RuleOrderWithComplexField_AddressInfo_city",
              "targetPath": "city",
            },
          },
        ],
      },
      "BillingAddress:adi1:doNotSolicit": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "adi1",
          "contextName": "BillingAddress",
          "fieldName": "doNotSolicit",
          "fieldPath": "doNotSolicit",
          "fieldType": "BOOLEAN",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "doNotSolicit",
                  "contextId": "adi1",
                  "contextName": "BillingAddress",
                  "newValue": true,
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "AddressInfo",
              "payloadtype": "DEFAULT",
              "ruleName": "RuleOrderWithComplexField_AddressInfo_doNotSolicit",
              "targetPath": "doNotSolicit",
            },
          },
        ],
      },
      "PartyRole:pr1:limit": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "pr1",
          "contextName": "PartyRole",
          "fieldName": "limit",
          "fieldPath": "limit",
          "fieldType": "DECIMAL",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "limit",
                  "contextId": "pr1",
                  "contextName": "PartyRole",
                  "newValue": 100,
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "PartyRole",
              "payloadtype": "DEFAULT",
              "ruleName": "RuleOrderWithComplexField_PartyRole_limit",
              "targetPath": "limit",
            },
          },
        ],
      },
      "Policy:0:policyNumber": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "0",
          "contextName": "Policy",
          "fieldName": "policyNumber",
          "fieldPath": "policyNumber",
          "fieldType": "STRING",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "policyNumber",
                  "contextId": "0",
                  "contextName": "Policy",
                  "newValue": "doNotSolicit is true",
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "Policy",
              "payloadtype": "DEFAULT",
              "ruleName": "RuleOrderWithComplexField_Policy_policyNumber",
              "targetPath": "policyNumber",
            },
          },
        ],
      },
      "RRCoverage:rr1:deductibleAmount": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "rr1",
          "contextName": "RRCoverage",
          "fieldName": "deductibleAmount",
          "fieldPath": "deductibleAmount",
          "fieldType": "DECIMAL",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "deductibleAmount",
                  "contextId": "rr1",
                  "contextName": "RRCoverage",
                  "newValue": 10,
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "RRCoverage",
              "payloadtype": "DEFAULT",
              "ruleName": "RuleOrderWithComplexField_AddressInfo_street",
              "targetPath": "deductibleAmount",
            },
          },
        ],
      },
      "RRCoverage:rr1:limitAmount": Object {
        "contextFieldInfo": DefaultContextFieldInfo {
          "contextId": "rr1",
          "contextName": "RRCoverage",
          "fieldName": "limitAmount",
          "fieldPath": "limitAmount",
          "fieldType": "DECIMAL",
        },
        "ruleResults": Array [
          Object {
            "conditionEvaluationResult": DefaultConditionEvaluationResult {
              "conditionEvaluation": "APPLICABLE",
              "error": undefined,
            },
            "kind": 1,
            "payloadResult": Object {
              "events": Array [
                ValueChangedEvent {
                  "attributeTarget": "limitAmount",
                  "contextId": "rr1",
                  "contextName": "RRCoverage",
                  "newValue": 10,
                  "previousValue": undefined,
                },
              ],
              "type": 8,
            },
            "ruleInfo": DefaultRuleInfo {
              "context": "RRCoverage",
              "payloadtype": "DEFAULT",
              "ruleName": "RuleOrderWithComplexField_RRCoverage_limitAmount",
              "targetPath": "limitAmount",
            },
          },
        ],
      },
    },
  },
  "fieldResults": Object {
    "BillingAddress:adi1:city": Object {
      "fieldType": "STRING",
      "id": "BillingAddress:adi1:city",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "city",
      "ruleResults": Array [],
    },
    "BillingAddress:adi1:doNotSolicit": Object {
      "fieldType": "BOOLEAN",
      "id": "BillingAddress:adi1:doNotSolicit",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "doNotSolicit",
      "ruleResults": Array [],
    },
    "PartyRole:pr1:limit": Object {
      "fieldType": "DECIMAL",
      "id": "PartyRole:pr1:limit",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "limit",
      "ruleResults": Array [],
    },
    "Policy:0:policyNumber": Object {
      "fieldType": "STRING",
      "id": "Policy:0:policyNumber",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "policyNumber",
      "ruleResults": Array [],
    },
    "RRCoverage:rr1:deductibleAmount": Object {
      "fieldType": "DECIMAL",
      "id": "RRCoverage:rr1:deductibleAmount",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "deductibleAmount",
      "ruleResults": Array [],
    },
    "RRCoverage:rr1:limitAmount": Object {
      "fieldType": "DECIMAL",
      "id": "RRCoverage:rr1:limitAmount",
      "info": Object {
        "getContextInstanceId": [Function],
        "getContextName": [Function],
      },
      "isDisabled": undefined,
      "isHidden": undefined,
      "resolvedTargetPath": "limitAmount",
      "ruleResults": Array [],
    },
  },
}
`;
